Safely checkout qubes-builder

DemiMarie opened the pull request:
Verifying that a git commit has a valid tag pointing to it is complex.
Instead, just checkout a specific SHA and let `make get-sources` do the
heavy lifting.

DemiMarie committed:
Safely checkout qubes-builder

Verifying that a git commit has a valid tag pointing to it is complex.
Instead, just checkout a specific SHA and let `make get-sources` do the
heavy lifting.

marmarek commented on build-infra/build-vm.sls:
```diff
@@ -206,29 +206,22 @@ github.com:
 
 {% for builder in builders_list %}
-{% set config_baseurl = salt['pillar.get']('build-infra:build-envs:' + env + ':builders-list:' + builder + ':config:repository:baseurl', 'https://github.com/QubesOS/qubes-') %}
+{% set config_baseurl = salt['pillar.get']('build-infra:build-envs:' + env + ':builders-list:' + builder + ':config:repository:baseurl', 'ssh://github.com/QubesOS/qubes-') %}
```
No. I don't want any more secrets available in the build vm. And definitely don't want it to have write access to some repositories (which with github permissions model would require creating yet another separate github account).

marmarek commented on build-infra/build-vm.sls:
```diff
@@ -206,29 +206,22 @@ github.com:
-    - cwd: {{ builder }}
+  cmd.script:
+    - source: salt://build-infra/safe-checkout-sha
+    - args: ['--', 'ssh://git@github.com/QubesOS/qubes-builder', {{ builder|yaml_encode }}, 'e8293887c39a5e39fe70d2e96564681a37ef8248']
```
Again, ssh -> https, but see above.

marmarek commented on build-infra/build-vm.sls:
```diff
@@ -206,29 +206,22 @@ github.com:
-  git.latest:
-    - name : https://github.com/QubesOS/qubes-builder
-    - target: {{ builder }}
-    - user: user
```
I see [`git.latest` has `rev`](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.git.html#salt.states.git.latest) which seems to do almost exactly what we need.
I don't see an option to skip this if it's already cloned (specifically: avoid rollback, and do not report a failure).

DemiMarie replied:
I am not sure if `rev` does what we want.  For example, I am not sure if it passes `--no-checkout` to `git clone`.

marmarek commented on build-infra/build-vm.sls:
```diff
@@ -206,29 +206,22 @@ github.com:
 {{builder}}-init:
   cmd.run:
-    - name: "BUILDERCONF= GIT_URL_builder=https://github.com/QubesOS/qubes-builder COMPONENTS=builder make get-sources"
+    - name: "BUILDERCONF= GIT_URL_builder=ssh://github.com/QubesOS/qubes-builder COMPONENTS=builder make get-sources"
```
No.

marmarek commented on build-infra/build-vm.sls:
```diff
@@ -206,29 +206,22 @@ github.com:
-    - name: git verify-tag --raw "$(git describe)" 2>&1 >/dev/null | grep '^\[GNUPG:\] TRUST_FULLY'
-    - cwd: {{ builder }}
+  cmd.script:
+    - source: salt://build-infra/safe-checkout-sha
```
There should be also `creates: {{ builder }}`, otherwise it may download (and rollback) already existing clones.

DemiMarie replied:
I was not aware of that feature, thanks!

marmarek commented on build-infra/safe-checkout-sha:
```diff
@@ -0,0 +1,22 @@
+trap '[ "$success" = yes ] || rm -rf -- "$directory"' EXIT
+[[ ${#sha} -eq 40 ]] || [[ ${#sha} -eq 64 ]] || usage
+[[ $sha =~ ^[0-9a-f]+$ ]] || usage
+if [[ -d "$directory" ]]; then
```
Missing `!`?

DemiMarie replied:
fixed

marmarek commented on build-infra/safe-checkout-sha:
```diff
@@ -0,0 +1,22 @@
+fi
+git -C "$directory" checkout "$sha"
+git update-ref FETCH_HEAD "$sha"
+git reset --hard "$sha"
```
Missing `-C` in those two.
And also, it would be nice to leave it on master branch (reset to `$sha`) instead of detached head.

DemiMarie replied:
Good catch on the working directory mixup.  The purpose of `git update-ref` is to leave it on the master branch.

DemiMarie committed:
Respond to code review from @marmarek:

- ssh:// → https://
- fix inverted logic in checkout script
- change directory for all git commands in checkout script
- mark checkout script as creating the builder directory

DemiMarie committed:
Check that builder names are valid

As they are interpolated into YAML with no escaping, a bad builder name
would be very bad.

marmarek commented on build-infra/build-vm.sls:
```diff
@@ -211,19 +217,12 @@ github.com:
-    - cwd: {{ builder }}
+  cmd.script:
+    - source: salt://build-infra/safe-checkout-sha
+    - args: ['--', 'https://git@github.com/QubesOS/qubes-builder', {{ builder|yaml_encode }}, 'e8293887c39a5e39fe70d2e96564681a37ef8248']
```
`git@` leftover from ssh

DemiMarie committed:
Tell Salt that the script creates the builder dir

DemiMarie committed:
Remove unneeded curly braces

DemiMarie committed:
Remove git@
